### DISCUSSION TITLE
Is it impossible to use one ORM-Mapped class to copy data from one database to another?

### DISCUSSION THREAD

**Discussion by jiseongg** [sqlalchemy/sqlalchemy#10822] [Usage Questions] [answered]
Hi, I am trying to use SQLAlchemy ORM to build an application for data management.

Me and my colleagues are considering the situation where we occasionally move data from one database to another. And with SQLAlchemy ORM, at first glance, we thought it can be easily achieved through the code like following:
```python
class User(DeclarativeBase):
    __tablename__ = "TB_EXAMPLE"
    id: Mapped[int] = mapped_column(primary_key=True)
    value: Mapped[int] = mapped_column()


src_engine = create_engine("~")
dst_engine = create_engine("~~")

with Session(src_engine) as session:
    result = session.execute(select(User))
    src_records = [r[0] for r in result]

User.metadata.create_all(dst_engine)
with Session(dst_engine) as session:
    session.bulk_save_objects(src_records)
    session.commit()
```

It didn't raise any exception, so I thought it succeeded. But in a database mapped to `dst_engine`, the table `TB_EXAMPLE` has been created but the data has not been added.

After searching communities, I found the thread [here](https://github.com/sqlalchemy/sqlalchemy/discussions/10761), and started to wonder whether the task I want to achieve is fundamentally impossible with SQLAlchemy ORM. Are there any ways to implement such functionality using SQLAlchemy ORM? 

Thanks!

**Comment by zzzeek** [comment, marked as answer]
when you load the objects they are in a state called [persistent](https://docs.sqlalchemy.org/en/20/glossary.html#term-persistent) which indicates they correspond to an already existing row in the database, and you'd get only an UPDATE when run into ORM unit-of-work functions of which bulk_save_objects() makes use of.

for your recipe to work as given you'd need to convert the objects to transient first using [make_transient](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.make_transient):

```py
from sqlalchemy.orm import make_transient
for obj in src_records:
    make_transient(obj)
```

then the unit of work would see them as non-inserted rows and do an INSERT.

you can work more directly on a per-table instance using `select()` and `insert()` statements using `Table` objects as the subjects also.

**Reply by jiseongg** [reply]
I had failed to find that function. Thank you for the information. :)
I succeeded to make that code work!

**Reply by zzzeek** [reply]
not a common function so no worries

**Reply by pmiddend** [reply]
I've tried this with an ORM object that has a `relationship` to another object and it seems `make_transient` does not work transitively? The error I'm getting is basically `Object <related-object> is already attached to session '2' (this is '1')`.

**Reply by zzzeek** [reply]
no make_transient doesnt do anything like that, it's entirely manual.  you can walk the object graph yourself if you wanted to use it that way.

The "easy" way to transfer ORM objects to another database is to use [session.merge()](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.merge).   this will follow along relationship paths and all that, it's just a generally inefficient method since it emits a SELECT for every row (i.e. it has no batching feature amongst many objects).

here's a demo that transfers an object + relationship from one database to another

```py
from __future__ import annotations

import os

from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import func
from sqlalchemy import select
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.orm import selectinload
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[str]
    bs: Mapped[list[B]] = relationship("B")


class B(Base):
    __tablename__ = "b"
    id: Mapped[int] = mapped_column(primary_key=True)
    a_id: Mapped[int] = mapped_column(ForeignKey("a.id"))
    data: Mapped[str]

if os.path.exists("db1.db"):
    os.unlink("db1.db")
if os.path.exists("db2.db"):
    os.unlink("db2.db")

e1 = create_engine("sqlite:///db1.db", echo=True)
Base.metadata.create_all(e1)

s1 = Session(e1)

s1.add(A(data="a1", bs=[B(data="b1"), B(data="b2"), B(data="b3")]))
s1.commit()



db1_as = s1.scalars(select(A).options(selectinload(A.bs)))


e2 = create_engine("sqlite:///db2.db", echo=True)

Base.metadata.create_all(e2)

s2 = Session(e2)

db2_as = [s2.merge(obj) for obj in db1_as]

s2.commit()

assert s2.scalar(select(func.count()).select_from(A)) == 1
assert s2.scalar(select(func.count()).select_from(B)) == 3
```
